File: backend/spotify/spotify.py
import time
import requests
from typing import Optional, Dict
from config import SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET, SPOTIFY_REDIRECT_URI

# Minimal Spotify helpers: client credentials and authorization code flows.
_token_cache = {"access_token": None, "expires_at": 0}

def _get_client_credentials_token() -> Optional[str]:
    if not SPOTIFY_CLIENT_ID or not SPOTIFY_CLIENT_SECRET:
        return None
    if _token_cache["access_token"] and _token_cache["expires_at"] > time.time():
        return _token_cache["access_token"]
    resp = requests.post(
        "https://accounts.spotify.com/api/token",
        data={"grant_type": "client_credentials"},
        auth=(SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET),
        timeout=10,
    )
    if resp.status_code != 200:
        return None
    data = resp.json()
    _token_cache["access_token"] = data["access_token"]
    _token_cache["expires_at"] = time.time() + data.get("expires_in", 3600) - 10
    return _token_cache["access_token"]

def search_track(q: str, limit: int = 5):
    token = _get_client_credentials_token()
    if not token:
        return []
    headers = {"Authorization": f"Bearer {token}"}
    params = {"q": q, "type": "track", "limit": limit}
    resp = requests.get("https://api.spotify.com/v1/search", headers=headers, params=params, timeout=10)
    if resp.status_code != 200:
        return []
    items = resp.json().get("tracks", {}).get("items", [])
    results = []
    for it in items:
        results.append({
            "spotify_id": it["id"],
            "title": it["name"],
            "artist": ", ".join([a["name"] for a in it.get("artists", [])]),
            "url": it["external_urls"].get("spotify"),
        })
    return results

def build_authorize_url(state: str, scopes: str = "user-read-private user-read-email playlist-modify-public") -> Optional[str]:
    """
    Return a Spotify authorization URL for the Authorization Code flow.
    state: opaque value your app uses to verify the response (server-side nonce)
    """
    if not SPOTIFY_CLIENT_ID:
        return None
    from urllib.parse import urlencode
    params = {
        "client_id": SPOTIFY_CLIENT_ID,
        "response_type": "code",
        "redirect_uri": SPOTIFY_REDIRECT_URI,
        "scope": scopes,
        "state": state,
        "show_dialog": "true",
    }
    url = "https://accounts.spotify.com/authorize?" + urlencode(params)
    return url

def exchange_code_for_token(code: str) -> Optional[Dict]:
    """
    Exchange an authorization code for access and refresh tokens.
    Returns dict with access_token, refresh_token, expires_in (seconds), or None.
    """
    if not SPOTIFY_CLIENT_ID or not SPOTIFY_CLIENT_SECRET:
        return None
    data = {
        "grant_type": "authorization_code",
        "code": code,
        "redirect_uri": SPOTIFY_REDIRECT_URI,
    }
    resp = requests.post(
        "https://accounts.spotify.com/api/token",
        data=data,
        auth=(SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET),
        timeout=10,
    )
    if resp.status_code != 200:
        return None
    return resp.json()

def refresh_access_token(refresh_token: str) -> Optional[Dict]:
    """
    Use a refresh token to obtain a new access token.
    Returns dict with access_token and expires_in (and possibly refresh_token), or None.
    """
    if not SPOTIFY_CLIENT_ID or not SPOTIFY_CLIENT_SECRET:
        return None
    data = {
        "grant_type": "refresh_token",
        "refresh_token": refresh_token,
    }
    resp = requests.post(
        "https://accounts.spotify.com/api/token",
        data=data,
        auth=(SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET),
        timeout=10,
    )
    if resp.status_code != 200:
        return None
    return resp.json()

File: services/adaptive/src/providers/spotify.ts
import type { Track } from "@biomusic/core";
import { loadEnv } from "../env.js";
import { getCache } from "../cache.js";
import type { MusicProvider, SearchParams } from "./index.js";

interface SpotifyTrack {
  id: string;
  name: string;
  duration_ms: number;
  preview_url: string | null;
  artists: Array<{ name: string }>;
  album: { name: string; images: Array<{ url: string }> };
}

/**
 * Spotify search via the client-credentials flow — app-level token, no
 * per-user OAuth required for catalogue search. Playback in the client still
 * uses the user's own Spotify session; this endpoint only surfaces candidates.
 */
export const spotifyProvider: MusicProvider = {
  id: "spotify",

  isConfigured() {
    const env = loadEnv();
    return Boolean(env.SPOTIFY_CLIENT_ID && env.SPOTIFY_CLIENT_SECRET);
  },

  async search(params: SearchParams): Promise<Track[]> {
    const token = await getAppToken();
    if (!token) return [];

    const url = new URL("https://api.spotify.com/v1/search");
    url.searchParams.set("type", "track");
    url.searchParams.set("limit", String(params.limit));
    url.searchParams.set("q", params.q?.trim() || "year:2020-2025");

    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
      signal: AbortSignal.timeout(7000),
    });
    if (!res.ok) return [];
    const data = (await res.json()) as { tracks?: { items?: SpotifyTrack[] } };
    return (data.tracks?.items ?? []).map((t) => ({
      title: t.name,
      artist: t.artists.map((a) => a.name).join(", "),
      album: t.album.name,
      durationMs: t.duration_ms,
      provider: "spotify" as const,
      providerTrackId: t.id,
      previewUrl: t.preview_url ?? undefined,
      artworkUrl: t.album.images[0]?.url,
    }));
  },
};

/** App access token, cached until shortly before expiry. */
async function getAppToken(): Promise<string | null> {
  const env = loadEnv();
  if (!env.SPOTIFY_CLIENT_ID || !env.SPOTIFY_CLIENT_SECRET) return null;

  const cached = await getCache().get<string>("spotify:app_token");
  if (cached) return cached;

  const basic = Buffer.from(`${env.SPOTIFY_CLIENT_ID}:${env.SPOTIFY_CLIENT_SECRET}`).toString("base64");
  const res = await fetch("https://accounts.spotify.com/api/token", {
    method: "POST",
    headers: { Authorization: `Basic ${basic}`, "Content-Type": "application/x-www-form-urlencoded" },
    body: "grant_type=client_credentials",
    signal: AbortSignal.timeout(7000),
  });
  if (!res.ok) return null;
  const json = (await res.json()) as { access_token?: string; expires_in?: number };
  if (!json.access_token) return null;
  await getCache().set("spotify:app_token", json.access_token, (json.expires_in ?? 3600) - 60);
  return json.access_token;
}
